### PR TITLE
Deploy more smart pointers in WebAutomationSession.cpp and various UIProcess classes

### PR DESCRIPTION
**Pull request by basuke**
#### 4c9c7334100c32116913c98a7e39d6ba553aa2dd
<pre>
Deploy more smart pointers in WebAutomationSession.cpp and various UIProcess classes
<a href="https://bugs.webkit.org/show_bug.cgi?id=279839">https://bugs.webkit.org/show_bug.cgi?id=279839</a>

Reviewed by Charlie Wolfe.

Deploy more smart pointers in WebAutomationSession.cpp as warned by the clang static analyzer.
Also make WebPageProxy.identifier() to be inlined to suppress static analyzer warnings.

* Source/WebCore/dom/Node.h:
(WebCore::Node::protectedTreeScope const):
* Source/WebCore/dom/TreeScope.h:
* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::WebAutomationSession::handleForWebFrameID):
(WebKit::WebAutomationSession::deleteAllCookies):
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::displayPropertiesChanged):
* Source/WebKit/UIProcess/Media/RemoteMediaSessionCoordinatorProxy.cpp:
(WebKit::RemoteMediaSessionCoordinatorProxy::RemoteMediaSessionCoordinatorProxy):
(WebKit::RemoteMediaSessionCoordinatorProxy::~RemoteMediaSessionCoordinatorProxy):
(WebKit::RemoteMediaSessionCoordinatorProxy::seekSessionToTime):
(WebKit::RemoteMediaSessionCoordinatorProxy::playSession):
(WebKit::RemoteMediaSessionCoordinatorProxy::pauseSession):
(WebKit::RemoteMediaSessionCoordinatorProxy::setSessionTrack):
(WebKit::RemoteMediaSessionCoordinatorProxy::coordinatorStateChanged):
(WebKit::RemoteMediaSessionCoordinatorProxy::protectedWebPageProxy):
(WebKit::m_logIdentifier): Deleted.
* Source/WebKit/UIProcess/Media/RemoteMediaSessionCoordinatorProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::Internals::Internals):
(WebKit::WebPageProxy::WebPageProxy):
(WebKit::WebPageProxy::~WebPageProxy):
(WebKit::WebPageProxy::initializeWebPage):
(WebKit::WebPageProxy::loadRequestWithNavigationShared):
(WebKit::WebPageProxy::loadFile):
(WebKit::WebPageProxy::didStartProvisionalLoadForFrameShared):
(WebKit::WebPageProxy::didFailProvisionalLoadForFrameShared):
(WebKit::WebPageProxy::didCommitLoadForFrame):
(WebKit::WebPageProxy::creationParameters):
(WebKit::WebPageProxy::identifier const): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxyInternals.h:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::removeWebPage):
(WebKit::WebProcessProxy::didFinishLaunching):
(WebKit::WebProcessProxy::canBeAddedToWebProcessCache const):
* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.cpp:
(WebKit::WebAutomationSessionProxy::getAccessibilityObjectForNode):
(WebKit::WebAutomationSessionProxy::ensureObserverForFrame):
(WebKit::WebAutomationSessionProxy::evaluateJavaScriptFunction):
(WebKit::WebAutomationSessionProxy::resolveChildFrameWithName):
(WebKit::WebAutomationSessionProxy::computeElementLayout):
(WebKit::WebAutomationSessionProxy::selectOptionElement):
(WebKit::WebAutomationSessionProxy::getCookiesForFrame):
(WebKit::WebAutomationSessionProxy::deleteCookie):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
(WebKit::WebPage::protectedCorePage const):

Canonical link: <a href="https://commits.webkit.org/283867@main">https://commits.webkit.org/283867@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/1d3a32543b91c10d654aa0b26382004dbfd66813

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/67613 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/46992 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/20245 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/71666 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/18750 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/69731 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/54790 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/18554 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/54149 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [loading-orange 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/12542 "Build is in progress. Recent messages:Printed configuration; Running apply-patch; Checked out pull request; Skipped layout-tests; layout-tests running") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/70680 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/43127 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/58477 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/34619 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/39800 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/15882 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/17112 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/61759 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/16224 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/73362 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/11572 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/15536 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/61593 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/11607 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/58545 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/61646 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14995 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/9453 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/3069 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/42798 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/43876 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/45065 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/43615 "Built successfully") | | |

### CODE REVIEW

File: Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp [hunk 1]
```diff
@@ -238,7 +238,7 @@ String WebAutomationSession::handleForWebFrameID(std::optional<FrameIdentifier>
     if (!frameID)
         return emptyString();
 
-    if (auto* frame = WebFrameProxy::webFrame(*frameID); frame && frame->isMainFrame())
+    if (RefPtr frame = WebFrameProxy::webFrame(*frameID); frame && frame->isMainFrame())
```
Comment (charliewolfe) [inline Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:244]: Why is this needed? `isMainFrame()` is trivial and inlined.

File: Source/WebKit/UIProcess/WebPageProxy.cpp [hunk 2]
```diff
@@ -1966,7 +1966,7 @@ void WebPageProxy::loadRequestWithNavigationShared(Ref<WebProcessProxy>&& proces
         if (!weakProcess->isLaunching() || !url.protocolIsFile())
             weakProcess->send(Messages::WebPage::LoadRequest(WTFMove(loadParameters)), webPageID);
         else
-            weakProcess->send(Messages::WebPage::LoadRequestWaitingForProcessLaunch(WTFMove(loadParameters), weakThis->internals().pageLoadState.resourceDirectoryURL(), weakThis->internals().identifier, true), webPageID);
+            weakProcess->send(Messages::WebPage::LoadRequestWaitingForProcessLaunch(WTFMove(loadParameters), weakThis->internals().pageLoadState.resourceDirectoryURL(), weakThis->identifier(), true), webPageID);
```
Comment (charliewolfe) [inline Source/WebKit/UIProcess/WebPageProxy.cpp]: I know you're not changing these, but `weakThis` and `weakProcess` should both be protected.

File: Source/WebKit/UIProcess/WebPageProxy.cpp [hunk 3]
```diff
@@ -2043,7 +2043,7 @@ RefPtr<API::Navigation> WebPageProxy::loadFile(const String& fileURLString, cons
 
         weakProcess->markProcessAsRecentlyUsed();
         if (weakProcess->isLaunching())
-            weakThis->send(Messages::WebPage::LoadRequestWaitingForProcessLaunch(WTFMove(loadParameters), resourceDirectoryURL, weakThis->internals().identifier, checkAssumedReadAccessToResourceURL));
+            weakThis->send(Messages::WebPage::LoadRequestWaitingForProcessLaunch(WTFMove(loadParameters), resourceDirectoryURL, weakThis->identifier(), checkAssumedReadAccessToResourceURL));
```
Comment (charliewolfe) [inline Source/WebKit/UIProcess/WebPageProxy.cpp]: Same thing here - `weakThis` and `weakProcess` need RefPtr

File: Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp (hunk 1 above)
Comment (basuke) [reply Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:244]: Based on this static analysis. https://build.webkit.org/results/Apple-Sonoma-Smart-Pointer-Static-Analyzer-Build/283793@main%20(1230)/scan-build-output/StaticAnalyzer/WebKit/StaticAnalyzerReports/report-WebAutomationSession.cpp-unknown-0-87952b.html#EndPath

If we don't need to check frame first, then this won't be the issue.

File: Source/WebKit/UIProcess/WebPageProxy.cpp (hunk 2 above)
Comment (basuke) [reply Source/WebKit/UIProcess/WebPageProxy.cpp]: We are focusing IPC related classes for now so that other part is not touched. This change is to settle many warnings in those focus area, WebPageProxy itself is very untouched in smart pointer context.

File: Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp (hunk 1 above)
Comment (charliewolfe) [reply Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:244]: Ok, but it doesn’t seem correct for the static analyzer to flag this because of a null check.

File: Source/WebKit/UIProcess/WebPageProxy.cpp (hunk 3 above)
Comment (charliewolfe) [reply Source/WebKit/UIProcess/WebPageProxy.cpp]: Also, why are we removing identifier from internals? `internals().identifier` seems safe.

File: Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp (hunk 1 above)
Comment (cdumez) [reply Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:244]: cc @rniwa : False positive?

Comment (rniwa) [reply Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:244]: So the issue here is that `WebFrameProxy::webFrame` is a non-trivial out-of-line function. Trivial function / statement checks currently can't handle a part of compound statement to be trivial and other parts to be non-trivial. The whole compound statement needs to be either entirely trivial or the whole thing will be deemed non-trivial. It can't perform data flow analysis either. So here, the fact `frame` is only used after the call to `WebFrameProxy::webFrame` isn't something static analyzer understands.

I suggest we either suppress the warning or deploy `RefPtr`.

Comment (cdumez) [reply Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:244]: It's not perf sensitive so I vote for RefPtr.

File: Source/WebKit/UIProcess/WebPageProxy.cpp (hunk 2 above)
Comment (cdumez) [reply Source/WebKit/UIProcess/WebPageProxy.cpp]: I don't think Charlie is asking for a lot here and his proposal is correct and would make the code safer, which is what you're working on. Therefore, I suggest we apply his suggestions.

Comment (basuke) [reply Source/WebKit/UIProcess/WebPageProxy.cpp]: Got it. I'm fixing visible part.

File: Source/WebKit/UIProcess/WebPageProxy.cpp (hunk 3 above)
Comment (basuke) [reply Source/WebKit/UIProcess/WebPageProxy.cpp]: I'm sorry I don't get the point maybe. I'm confusing why double dereference is safe. Please fix my thinking. WebPage's internal is deferent memory block and needs to dereference twice. It cannot fit in one instruction so that it should be treaded as not safe. This is my understanding.

**Review by charliewolfe — APPROVED** [review]

**Comment by basuke** [comment]
@charliewolfe Thanks!
